src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/UnixPkcs12Reader.cs

```diff
@@ -0,0 +1,670 @@
+            // The next most common PFX contains 3 certs, and one key.
+            //
+            // Nothing requires that there be fewer keys than certs,
+            // but it's sort of nonsensical when loading this way.
+            CertBagAsn[] certBags = ArrayPool<CertBagAsn>.Shared.Rent(10);
+            AttributeAsn[][] certBagAttrs = ArrayPool<AttributeAsn[]>.Shared.Rent(10);
+            SafeBagAsn[] keyBags = ArrayPool<SafeBagAsn>.Shared.Rent(10);
+            SubjectPublicKeyInfoAsn[] publicKeyInfos = null;
+            AsymmetricAlgorithm[] keys = null;
+            int certBagIdx = 0;
+            int keyBagIdx = 0;
+
+            try
+            {
+                for (int i = 0; i < _safeContentsValues.Length; i++)
+                {
+                    string contentType = _safeContentsValues[i].ContentType;
+                    bool process = false;
+
+                    if (contentType == null)
+                    {
+                        break;
+                    }
+
+                    // Should enveloped throw here?
+                    if (contentType == Oids.Pkcs7Data)
+                    {
+                        process = true;
+                    }
+                    else if (contentType == Oids.Pkcs7Encrypted)
+                    {
+                        DecryptSafeContents(password, ref _safeContentsValues[i]);
+                        process = true;
+                    }
+
+                    if (process)
+                    {
+                        ProcessSafeContents(
+                            _safeContentsValues[i],
+                            ref certBags,
+                            ref certBagAttrs,
+                            ref certBagIdx,
+                            ref keyBags,
+                            ref keyBagIdx);
+                    }
+                }
+
+                _certs = ArrayPool<CertAndKey>.Shared.Rent(certBagIdx);
+                _certs.AsSpan().Clear();
+
+                keys = ArrayPool<AsymmetricAlgorithm>.Shared.Rent(keyBagIdx);
+                keys.AsSpan().Clear();
+
+                publicKeyInfos = ArrayPool<SubjectPublicKeyInfoAsn>.Shared.Rent(keyBagIdx);
+                publicKeyInfos.AsSpan().Clear();
+
+                byte[] spkiBuf = null;
+
+                for (int i = keyBagIdx - 1; i >= 0; i--)
+                {
+                    byte[] publicKeyBytes = null;
+
+                    try
+                    {
+                        SafeBagAsn keyBag = keyBags[i];
+                        AsymmetricAlgorithm key = LoadKey(keyBag, password);
+                        publicKeyBytes = CryptoPool.Rent(keyBag.BagValue.Length);
+
+                        int pubLength;
+
+                        while (!key.TryExportSubjectPublicKeyInfo(spkiBuf, out pubLength))
+                        {
+                            byte[] toReturn = spkiBuf;
+                            spkiBuf = CryptoPool.Rent((toReturn?.Length ?? 128) * 2);
```

Comment:
why is public key using crypto pool vs regular pool if it doesn't have to be cleared?

Reply:
If CryptoPool was actually a separate pool, it would avoid a double-returned array (therefore double-rentable) having an influence on this match.

In the crypto code, the answer is "always use CryptoPool (for byte[]), unless you're sending the buffer as a return or into a callback that can leave the crypto stack, then don't use it".